File: cross-language-features.ml
module Xl = Willyhorizont.Runtime

let () =
    (*
    1. support closure as value, or has workaround
    *)
    let say_hello = Xl.Type (fun va -> (
        let itr = Seq.to_dispenser (List.to_seq (Xl.to_list va)) in
        let callback_function = Xl.next itr in
        print_endline "hello";
        (Xl.to_closure callback_function) (Xl.Type [Xl.Type ()])
    )) in
    let _ = Xl.Type ((Xl.to_closure say_hello) (Xl.Type [Xl.Type (fun (_) -> (
        print_endline "world";
        Xl.Type ()
    ))])) in
    (* let create_multiplier = Xl.Type (fun va -> (
        let itr = Seq.to_dispenser (List.to_seq (Xl.to_list va)) in
        let aa = Xl.next itr in
        Xl.Type (fun va -> (
            let itr = Seq.to_dispenser (List.to_seq (Xl.to_list va)) in
            let bb = Xl.next itr in
            Xl.Type ((Xl.to_int aa) * (Xl.to_int bb))
        ))
    )) in
    let multiply_by_two = Xl.Type ((Xl.to_closure create_multiplier) (Xl.Type [Xl.Type (Xl.to_int 2)])) in
    let () = print_endline ("multiply_by_two(10): " ^ (Xl.to_string (Xl.json_stringify (Xl.Type [Xl.Type ((Xl.to_closure multiply_by_two) (Xl.to_int 10))])))) in *)
    (*
    2. support dynamic-typed value, or has workaround
    *)
    let some_py_list = Xl.Type [
        Xl.Type ();
        Xl.Type true;
        Xl.Type false;
        Xl.Type "foo";
        Xl.Type 0;
        Xl.Type (-123);
        Xl.Type (123.789);
        Xl.Type (-123.789);
        Xl.Type [Xl.Type 1; Xl.Type 2; Xl.Type 3];
        Xl.Type [Xl.Type [Xl.Type "foo"; Xl.Type "bar"]];
        Xl.Type (fun va -> (
            let itr = Seq.to_dispenser (List.to_seq (Xl.to_list va)) in
            let aa = Xl.next itr in
            let bb = Xl.next itr in
            Xl.Type ((Xl.to_int aa) * (Xl.to_int bb))
        ));
    ] in
    let () = print_endline ("some_py_list: " ^ (Xl.to_string (Xl.json_stringify (Xl.Type [some_py_list])))) in
    let () = print_endline ("some_py_list: " ^ (Xl.to_string (Xl.json_stringify (Xl.Type [some_py_list; Xl.Type [Xl.Type [Xl.Type "pretty"; Xl.Type true]]])))) in
    let some_py_dict = Xl.Type [
        Xl.Type [Xl.Type "some_py_none"; Xl.Type ()];
        Xl.Type [Xl.Type "some_py_boolean_true"; Xl.Type true];
        Xl.Type [Xl.Type "some_py_boolean_false"; Xl.Type false];
        Xl.Type [Xl.Type "some_js_string"; Xl.Type "foo"];
        Xl.Type [Xl.Type "some_js_int_positive"; Xl.Type 0];
        Xl.Type [Xl.Type "some_js_int_negative"; Xl.Type (-123)];
        Xl.Type [Xl.Type "some_js_float_positive"; Xl.Type (123.789)];
        Xl.Type [Xl.Type "some_js_float_negative"; Xl.Type (-123.789)];
        Xl.Type [Xl.Type "some_py_list"; Xl.Type [Xl.Type 1; Xl.Type 2; Xl.Type 3]];
        Xl.Type [Xl.Type "some_py_dict"; Xl.Type [Xl.Type [Xl.Type "foo"; Xl.Type "bar"]]];
        Xl.Type [Xl.Type "some_js_function"; Xl.Type (fun va -> (
            let itr = Seq.to_dispenser (List.to_seq (Xl.to_list va)) in
            let aa = Xl.next itr in
            let bb = Xl.next itr in
            Xl.Type ((Xl.to_int aa) * (Xl.to_int bb))
        ))];
    ] in
    let () = print_endline ("some_py_dict: " ^ (Xl.to_string (Xl.json_stringify (Xl.Type [some_py_dict])))) in
    let () = print_endline ("some_py_dict: " ^ (Xl.to_string (Xl.json_stringify (Xl.Type [some_py_dict; Xl.Type [Xl.Type [Xl.Type "pretty"; Xl.Type true]]])))) in
    ()

File: languages/visual-basic-dot-net/cross-language-features.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports WillyHorizont.Runtime.Xl

Module Program
    Sub Main()
        ' 1. support lambda as value, or has workaround
        Dim SayHello As Object = Xl.InitLambda(Function(Va)
            Dim Itr As Object = Xl.Iter(Va)
            Dim Callback As Object = Xl.GetNext(Itr)
            Console.WriteLine("hello")
            Callback.Invoke()
        End Function)
        SayHello.Invoke(Xl.InitLambda(Function(Va)
            Console.WriteLine("world")
        End Function))
        Dim CreateMultiplier As Object = Xl.InitLambda(Function(VaAa)
            Dim ItrAa As Object = Xl.Iter(VaAa)
            Dim Aa As Object = Xl.GetNext(ItrAa)
            Return Xl.InitLambda(Function(VaBb)
                Dim ItrBb As Object = Xl.Iter(VaBb)
                Dim Bb As Object = Xl.GetNext(ItrBb)
                Return Aa * Bb
            End Function)
        End Function)
        Dim MultiplyByTwo As Object = CreateMultiplier.Invoke(2)
        Console.WriteLine($"multiply_by_two(10): {MultiplyByTwo.Invoke(10)}")
        Dim MultiplyByEight As Object = CreateMultiplier.Invoke(8)
        Console.WriteLine($"multiply_by_eight(4): {MultiplyByEight.Invoke(4)}")
        Console.WriteLine($"multiply_by_two(8): {MultiplyByTwo.Invoke(8)}")

        ' 2. support dynamic-typed value, or has workaround
        Dim XlList As Object = Xl.InitList(
            Nothing,
            True,
            False,
            "foo",
            0,
            -123,
            123.789,
            -123.789,
            Xl.InitList(1, 2, 3),
            Xl.InitDict(Xl.InitPair("foo", "bar")),
            Xl.InitLambda(Function(Va)
                Dim Itr As Object = Xl.Iter(Va)
                Dim Aa As Object = Xl.GetNext(Itr)
                Dim Bb As Object = Xl.GetNext(Itr)
                Return Aa * Bb
            End Function)
        )
        Console.WriteLine($"xl_list: {Xl.JsonStringify(XlList)}")
        Console.WriteLine($"xl_list: {Xl.JsonStringify(XlList, Pretty:=True)}")
        Dim XlDict As Object = Xl.InitDict(
            Xl.InitPair("xl_none", Nothing),
            Xl.InitPair("xl_bool_true", True),
            Xl.InitPair("xl_bool_false", False),
            Xl.InitPair("xl_string", "foo"),
            Xl.InitPair("xl_int_positive", 0),
            Xl.InitPair("xl_int_negative", -123),
            Xl.InitPair("xl_float_positive", 123.789),
            Xl.InitPair("xl_float_negative", -123.789),
            Xl.InitPair("xl_list", Xl.InitList(1, 2, 3)),
            Xl.InitPair("xl_dict", Xl.InitDict(Xl.InitPair("foo", "bar"))),
            Xl.InitPair("xl_lambda", Xl.InitLambda(Function(Va)
                Dim Itr As Object = Xl.Iter(Va)
                Dim Aa As Object = Xl.GetNext(Itr)
                Dim Bb As Object = Xl.GetNext(Itr)
                Return Aa * Bb
            End Function))
        )
        Console.WriteLine($"xl_dict: {Xl.JsonStringify(XlDict)}")
        Console.WriteLine($"xl_dict: {Xl.JsonStringify(XlDict, Pretty:=True)}")
    End Sub
End Module
